### PR TITLE
Implement conditional parameter types

### PR DESCRIPTION
**Pull request by rvanvelzen**
Resolves phpstan/phpstan#7124. Parameter types could in fact already be conditional when using template types, this just implements the same logic as used for return types.

To improve error messages for conditional types in general they're resolved immediately now (when possible)

### CODE REVIEW

File: tests/PHPStan/Analyser/AnalyserIntegrationTest.php [hunk 1]
```diff
@@ -768,7 +768,24 @@ public function testBug7135(): void
...
+	}
+
+	public function testDiscussion7124(): void
+	{
+		if (PHP_VERSION_ID < 80000) {
+			$this->markTestSkipped('Test requires PHP 8.0.');
+		}
+
+		$errors = $this->runAnalyse(__DIR__ . '/data/discussion-7124.php');
+		$this->assertCount(4, $errors);
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/AnalyserIntegrationTest.php:792]: Please assert which errors are happening here.

File: tests/PHPStan/Analyser/AnalyserIntegrationTest.php [hunk 2]
```diff
@@ -768,7 +768,24 @@ public function testBug7135(): void
...
 		$this->assertCount(1, $errors);
-		$this->assertSame('Cannot create callable from the new operator.', $errors[0]->getMessage());
+		$this->assertSame('Parameter #2 $callback of function Discussion7124\filter expects callable(bool, int): bool, Closure(int, bool): bool given', $errors[0]->getMessage());
+		$this->assertSame(38, $errors[0]->getMessage());
+		$this->assertSame('Parameter #2 $callback of function Discussion7124\filter expects callable(bool, int): bool, Closure(int): bool given', $errors[0]->getMessage());
+		$this->assertSame(45, $errors[0]->getMessage());
+		$this->assertSame('Parameter #2 $callback of function Discussion7124\filter expects callable(int): bool, Closure(bool): bool given', $errors[0]->getMessage());
+		$this->assertSame(52, $errors[0]->getMessage());
+		$this->assertSame('Parameter #2 $callback of function Discussion7124\filter expects callable(bool): bool, Closure(int): bool given', $errors[0]->getMessage());
+		$this->assertSame(59, $errors[0]->getMessage());
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/AnalyserIntegrationTest.php]: What? :) You're asserting the same thing to different values.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: tests/PHPStan/Analyser/AnalyserIntegrationTest.php (hunk 2 above)
Comment (rvanvelzen) [reply tests/PHPStan/Analyser/AnalyserIntegrationTest.php]: I think I was day dreaming a bit when moving some code around :sweat_smile:

**Comment by rvanvelzen** [comment]
I've refactored "late resolvable types" a bit to make it more future-proof. Stuff like offset access types can reuse much of the logic.

File: build/composer-require-checker.json [hunk 3]
```diff
@@ -19,6 +19,7 @@
     "Reflection",
     "SPL",
     "standard",
+    "pcntl",
```
Comment (staabm) [inline build/composer-require-checker.json]: Did you re-add this stuff intentionally or was it a bad merge/rebase?

Comment (rvanvelzen) [reply build/composer-require-checker.json]: Definitely a bad merge

**Comment by ondrejmirtes** [comment]
Nice cleanup, I really like it :) Thank you!

**Review by ondrejmirtes — APPROVED** [review]

**Pull request merged**